src/charm.py

```diff
@@ -236,64 +235,58 @@ def _on_db_router_relation_changed(self, event: RelationChangedEvent) -> None:
         event_relation_databag["db_host"] = json.dumps(unit_address)
 
         application_unit = list(event.relation.units)[0]
-        application_data = copy.deepcopy(event.relation.data[application_unit])
-        application_info = collections.defaultdict(dict)
-
-        # Create the mysql router user if it does not already exist
-        try:
-            mysqlrouter_username = application_data.get("mysqlrouter_username")
-            mysqlrouter_user_exists = self._mysql.does_mysql_user_exist(mysqlrouter_username)
-
-            if mysqlrouter_username and not mysqlrouter_user_exists:
-                mysqlrouter_password = generate_random_password(PASSWORD_LENGTH)
-                self._mysql.configure_mysqlrouter_user(mysqlrouter_username, mysqlrouter_password)
-
-                event_relation_databag["mysqlrouter_password"] = json.dumps(mysqlrouter_password)
+        application_data = event.relation.data[application_unit]
 
-            if mysqlrouter_username:
-                event_relation_databag["mysqlrouter_allowed_units"] = json.dumps(unit_names)
-        except (MySQLCheckUserExistenceError, MySQLConfigureRouterUserError):
-            self.unit.status = BlockedStatus("Failed to create mysqlrouter user")
-            return
+        # Retrieve application names in the relation databag (which correspond to usernames)
+        application_names = set(
+            [key.split("_")[0] for key in application_data if "username" == key.split("_")[1]]
+        )
 
-        # Retrieve and compile application usernames and database names
-        for key, value in application_data.items():
-            application_name = key.split("_")[0]
+        for application_name in application_names:
+            username = application_data.get(f"{application_name}_username")
+            database = application_data.get(f"{application_name}_database")
 
-            if application_name == "mysqlrouter":
+            if not username or (application_name != "mysqlrouter" and not database):
+                logger.warning(
+                    f"Missing information for application {application_name} to create a database and scoped user"
+                )
                 continue
 
-            if "username" in key:
-                application_info[application_name]["username"] = value
+            try:
+                password = generate_random_password(PASSWORD_LENGTH)
 
-            if "database" in key:
-                application_info[application_name]["database"] = value
+                # Bootstrap the mysql router user
+                if application_name == "mysqlrouter":
+                    mysqlrouter_user_exists = self._mysql.does_mysql_user_exist(username)
+                    if not mysqlrouter_user_exists:
```

Comment:
If we find this variable before checking the application_name, we can use it on line 271